ping/ping_output.h

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright (c) 2024 Georg Pfuetzenreuter <mail+ip@georg-pfuetzenreuter.net>
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ * 3. Neither the name of the copyright holder nor the names of its
+ *    contributors may be used to endorse or promote products derived from
+ *    this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/* extract first argument whilst staying C99 compliant */
+#define _JSON_VALUE(value, ...) value
+#define JSON_VALUE(...) _JSON_VALUE(__VA_ARGS__, NULL)
+
+#define PRINT(msg, json_key, json_type, ...) \
+	do { \
+		if (rts->opt_json) \
+			construct_json(rts, json_type, json_key, JSON_VALUE(__VA_ARGS__)); \
+		else \
+			printf(msg, __VA_ARGS__); \
+	} while (0)
+#define PRINT_INT(msg, json_key, ...) PRINT(msg, json_key, PING_JSON_INT, __VA_ARGS__)
+#define PRINT_STR(msg, json_key, ...) PRINT(msg, json_key, PING_JSON_STR, __VA_ARGS__)
+
+#define ERRORF(status, errnum, ...) \
```

Comment:
I'm not happy with macro names (not understand what they mean), but let's leave it now. I would also not create new headers but put code into `ping/ping.h`.

Reply:
I understand, some are ambiguous too. I split it into a separate header file because ping.h is already quite large, but if you prefer I can merge it back in?

Reply:
It does not look important enough to have it separated. In the end JSON support would get into other tools (at least to arping), then it would go to `iputils_common.h`.

But it's not important, C code itself matters more (+ good names for macros and reconsider if `static inline` would be more readable instead of macros). Next week after LTP release I'll have a deeper look and ask @metan-ucw to do review as well.

Reply:
First of all, why are there such horrid macros to begin with? We can have just a regular functions such as 'void print_int(const char *msg, const char *key, int val)' instead. These can be static inline as well as @pevik suggested, or we can put them into ping_output.c. The biggest problem with macros as these is that we loose all type checking.

Reply:
I understand, and agree functions would be better suited; in an earlier comment both variants were suggested, I opted for the macro one, but it turned out to require multiple hoops.
Will try to refactor.